runtime/src/main/java/org/corfudb/util/ReflectionUtils.java

```diff
@@ -92,38 +100,94 @@ public static String getMethodNameOnlyFromString(String s) {
                 .toArray(Class[]::new);
     }
 
-    public static <T> T newInstanceFromUnknownArgumentTypes(Class<T> cls, Object[] args) {
-        try {
-            return cls.getDeclaredConstructor(getArgumentTypesFromArgumentList(args))
-                    .newInstance(args);
-        } catch (InstantiationException | InvocationTargetException | IllegalAccessException ie) {
-            throw new RuntimeException(ie);
-        } catch (NoSuchMethodException nsme) {
-            // Now we need to find a matching primitive type.
-            // We can maybe cheat by running through all the constructors until we get what we want
-            // due to autoboxing
-            Constructor[] ctors = Arrays.stream(cls.getDeclaredConstructors())
-                    .filter(c -> c.getParameterTypes().length == args.length)
-                    .toArray(Constructor[]::new);
-            for (Constructor<?> ctor : ctors) {
-                try {
-                    return (T) ctor.newInstance(args);
-                } catch (Exception e) {
-                    // silently drop exceptions since we are brute forcing here...
-                }
+    /**
+     * Given the constructor arguments and the desired class type, find
+     * the closest matching constructor. For example, if a wrapper
+     * class contains two constructors, Constructor(A) and Constructor(B),
+     * and the constructor argument is of type B, where B inherits from A,
+     * instantiate the object via Constructor(B) since it is closes with
+     * respect to type hierarchy.
+     *
+     * @param constructors   Object constructors.
+     * @param args           Constructor arguments.
+     * @param <T>            Type
+     * @return instantiated  wrapper class.
+     *
+     * @throws IllegalAccessException should not be thrown
+     * @throws InvocationTargetException should not be thrown
+     * @throws InstantiationException should not be thrown
+     */
+    public static <T> Object findMatchingConstructor(
+            @NonNull Constructor[] constructors, @NonNull Object[] args)
+            throws IllegalAccessException, InvocationTargetException, InstantiationException {
+        // Figure out the argument types.
+        final List<Class> argTypes = Arrays
+                .stream(args).map(Object::getClass)
+                .collect(Collectors.toList());
+        // Filter out the constructors that do not have the same arity.
+        final List<Constructor> constructorCandidates = Arrays.stream(constructors)
+                .filter(constructor -> constructor.getParameterTypes().length == args.length)
+                .collect(Collectors.toList());
```

Comment:
Btw this breaks support for non-static nested classes, since `getParameterTypes()` will return the enclosing instance as the first parameter, so all the constructors will be filtered out because the sizes won't match. 

From the javadoc:

> The parameterTypes parameter is an array of Class objects that identify the constructor's formal parameter types, in declared order. If this Class object represents an inner class declared in a non-static context, the formal parameter types include the explicit enclosing instance as the first parameter.

Actually, if you remove static modifier from the test types in `ObjectBuilderTest.java` the test will fail.

Reply:
Good point. This is a pretty minor use-case, which we do not have in Corfu right now. Maybe we can address this in a subsequent patch?

The plan was never to implement the specification fully, but rather improve on the previous approach.